src/plots/cartesian/autorange.js

```diff
@@ -226,8 +226,18 @@ function makePadFn(fullLayout, ax, max) {
     var A = 0;
     var B = 0;
     if(!isLinked(fullLayout, ax._id)) {
-        A = padInsideLabelsOnAnchorAxis(ax, max);
-        B = padInsideLabelsOnThisAxis(ax, max);
+        if(
+            ax.rangemode !== 'tozero' &&
+            ax.rangemode !== 'nonnegative'
+        ) {
+            A = padInsideLabelsOnAnchorAxis(ax, max);
+            if(
+                anchorAxis.rangemode !== 'tozero' &&
+                anchorAxis.rangemode !== 'nonnegative'
+            ) {
```

Comment:
I don't think we want any of these `rangemode` conditions, now that you have `nopad` in the fallback conditions above. Basically in your new `ticklabelposition-5` mock, only the blue one should have labels overlapping bars. Maybe you want to flip either the red or green labels to the bottom, in which case those would overlap the bars too, but otherwise just the blue should hit this condition.

Reply:
I removed one of the checks in https://github.com/plotly/plotly.js/pull/5331/commits/0b254e7667dc4f2656f0700576fc6b98bd409cef but kept the other one.

Reply:
The test images look good now, but I don't understand what the remaining check is doing - can you explain?

Reply:
I drop the second check as you suggested.
After further investigation, it looks now that we could possibly drop `padInsideLabelsOnThisAxis` function which used to  tweaks the range for giving the start/end labels more chance to appear on the graph. But I'd rather trying that in a separate PR.